dali/operators/image/clahe/clahe_op.cu

```diff
@@ -0,0 +1,1207 @@
+
+  const int BLOCK_SIZE = 128;
+  int block_start = blockIdx.x * BLOCK_SIZE;
+  int tid = threadIdx.x;
+  int N = H * W;
+
+  // Coalesced load: Each thread loads consecutive bytes
+  // This achieves 100% memory bus utilization vs 25% in naive version
+  if (block_start + tid < N && tid < BLOCK_SIZE) {
+    int global_idx = block_start + tid;
+    int rgb_base = global_idx * 3;
+
+    // Load RGB triplet (still somewhat strided, but better with caching)
+    s_rgb[0][tid] = rgb[rgb_base + 0];  // R
+    s_rgb[1][tid] = rgb[rgb_base + 1];  // G
+    s_rgb[2][tid] = rgb[rgb_base + 2];  // B
+  }
+  __syncthreads();
+
+  // Process from shared memory (no global memory access penalty)
+  if (block_start + tid < N && tid < BLOCK_SIZE) {
+    uint8_t r = s_rgb[0][tid];
+    uint8_t g = s_rgb[1][tid];
+    uint8_t b = s_rgb[2][tid];
+
+    // Convert to LAB L* to match OpenCV CLAHE behavior
+    float L, a, b_lab;
+    rgb_to_lab(r, g, b, &L, &a, &b_lab);
+
+    // Scale L [0,100] to [0,255] for consistency
+    uint8_t L_u8 = (uint8_t)lrintf(dali::clamp(L * 255.0f / 100.0f, 0.f, 255.f));
+    y_out[block_start + tid] = L_u8;
+  }
+}
+
+// Original version (fallback for small images)
+__global__ void rgb_to_y_u8_nhwc_kernel(const uint8_t *__restrict__ rgb,
+                                        uint8_t *__restrict__ y_out, int H, int W) {
+  int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  int N = H * W;
+  if (idx >= N) {
+    return;
+  }
+
+  int c0 = 3 * idx;
+  uint8_t r = rgb[c0 + 0];
+  uint8_t g = rgb[c0 + 1];
+  uint8_t b = rgb[c0 + 2];
+
+  // Convert to LAB L* to match OpenCV CLAHE behavior
+  float L, a, b_lab;
+  rgb_to_lab(r, g, b, &L, &a, &b_lab);
+
+  // Scale L [0,100] to [0,255] for consistency
+  uint8_t L_u8 = (uint8_t)lrintf(dali::clamp(L * 255.0f / 100.0f, 0.f, 255.f));
+  y_out[idx] = L_u8;
+}
+
+// Vectorized version for better memory bandwidth (processes 4 pixels at once)
+__global__ void rgb_to_y_u8_nhwc_vectorized_kernel(const uint8_t *__restrict__ rgb,
+                                                   uint8_t *__restrict__ y_out, int H, int W) {
+  int base_idx = (blockIdx.x * blockDim.x + threadIdx.x) * 4;
+  int N = H * W;
+
+#pragma unroll
+  for (int i = 0; i < 4; ++i) {
+    int idx = base_idx + i;
+    if (idx >= N) {
+      return;
+    }
+
+    int c0 = 3 * idx;
+
+    uint8_t r = rgb[c0 + 0];
+    uint8_t g = rgb[c0 + 1];
+    uint8_t b = rgb[c0 + 2];
+
+    float L, a, b_lab;
+    rgb_to_lab(r, g, b, &L, &a, &b_lab);
+
+    uint8_t L_u8 = (uint8_t)lrintf(dali::clamp(L * 255.0f / 100.0f, 0.f, 255.f));
+    y_out[idx] = L_u8;
+  }
+}
+
+// -------------------------------------------------------------------------------------
+// Histogram clipping, redistribution, and CDF calculation helper
+// -------------------------------------------------------------------------------------
+
+__device__ void clip_redistribute_cdf(unsigned int *h, int bins, int area, float clip_limit_rel,
```

Comment:
Optimizing this piece is likely out of scope of this initial PR, so please mark it with some `TODO` that this code needs some attention. Currently it has a lot of sequential computations involving global memory (lut) and it's amenable to parallelization, at least at warp level.